Replace Lodash with Ramda/Rambda

karlos1337 opened the pull request:
I'm not sure about adding both libraries instead of setting just one, but, for sure, I think would be preferable to use one of those instead of Lodash since are side-effects free and I've already found a lot of bugs in Lodash based projects on mutating structures by error.

davidgg commented:
Since we're using lodash in a functional way (that's our goal), IMO replacing it with ramda it's a good decision.

We all have to agree to approve this change so let the rest of the people speak @MarsBased/marsbased (specially @davidrivers93 @ItsGenis @xredo @javierartero ). This is not about Angular only but JS, so I'd like to hear @danigb 's opinion on that too and our collaborator @fernandomeridamatilla

davidrivers93 reviewed (approved):
I've not used a lot Rambda but I agree on using it instead of Lodash.

ItsGenis reviewed (approved):
We use Ramda at Rundit and it is quite cool. It has an arguable steep learning curve if you're not much into functional but I'm all in for it.

danigb commented:
I've used to like Rambda a lot in the past, but not anymore. Here are the reasons:

- Ramda doesn't play well with typescript. As Genís said, it's used extensively at Rundit, but we're discussing about abandon it because this reason.
- Ramda promotes a dot-free style of coding. While this style has it's advantages (and it's beautiful), I think most of the JS developers are not used to it, and feels weird.
- For the reason above, some things looks (are) unnecessary complex. I've seen some expressions at Rundit that could be expressed in a simpler way with plain JS. When Ramda was created some JS features (like arrow functions) were not very spread. Arrow functions are an easy and concise way to express lot of things that comes with Ramda.
- Compared to other utility libraries, Ramda has a steep learning curve

So, for those reasons I would think long and hard before adding Ramda to a project. 

Carlos mentioned the "mutation structures by error". I don't think Ramda solves this problem (I think if a developer makes that kind of mistake, will make more mistakes with Ramda). There are some other libraries that can be used to enforce mutation free changes (like https://github.com/immerjs/immer)

karlos1337 commented:
I totally agree with that reasons, in fact, I prefer the other alternative I've added to the doc called [Rambda](https://github.com/selfrefactor/rambda), which is a lightweight and simpler version of Ramda that also have better TS support and performance.
I added both to the doc because Ramba is more known, has better docs and I expect that some of you have already worked with it and would prefer it over Rambda.

Using native JS and removing Lodash at all is also an option but even that has some learning curve (restructuring, reducing, which Array methods could or not could be used, Object.keys/values/entries...) and resultant code is harder to understand than using any library.

Related to Immer, implies wrapping each operation in a `produce` function that could add a lot of verbosity, could be fine for being used [as an integration](https://datorama.github.io/akita/docs/immer), but even with that it doesn't supply utils for interacting with collections so will still need Lodash (or native JS) and the workaround will be something similar to adding inmutability to non-inmutable operations that seems wrong imo.

Anyway, the intention is not to add Ramda at any cost but to replace Lodash with a similar library that offer also immutability avoiding side effects, and also for taking advantage of it and being used in the state manager library in which immutable operations are a requirement. 
If something like Imutable-Lodash exists would perfectly suits for me but the closest libraries I've founded are Ramda and Rambda.

xredo commented:
I have not used Ramda extensively in the past (mostly because its learning curve) and I would like your help to give my opinion. My main concern is switching to Ramda (or Rambda) and then missing Lodash for some utilities. And ending up using both.

I would like to see a list of the methods that are commonly used in Lodash that are not supported by Ramda or ES6. Having the opposite (things widely used in Ramda that are not supported in ES6 or Lodash) will be a nice bonus.

Then, we'll still need to discuss the other points that you exposed as they are quite important.

davidgg commented:
To be honest, I prefer vanilla JS when code readability and size is not compromised. 

Since we're not aligned, and all are valid alternatives, what about adding both (or removing them from the guide) and let the developer decide which library use:

- During the setup of the project.
- When it's needed a library and there isn't one added to the project.

The rest of the time we should stick to the added library.

davidgg commented:
We've decided to remove them, and add a statement to make it clear that adding a new library should be approved by the tech lead (they will decide if lodash, ramda or rambda is the best alternative for the project)

davidgg commented on guides/development/angular-guidelines.md:
```diff
@@ -485,7 +485,7 @@ Use [Transloco](https://github.com/ngneat/transloco).
 - [Jest](https://jestjs.io/). General purpose JS testing.
 - [ESLint](https://eslint.org/). JS Linter
-- [Lodash](https://lodash.com/). Provides functions to manipulate data structures.
+- [Ramda](https://ramdajs.com)/[Rambda](https://github.com/selfrefactor/rambda). Provides immutability and side-effect free functions to manipulate data structures.
```
```suggestion
```

davidgg commented on guides/development/angular-guidelines.md:
```diff
@@ -514,4 +514,4 @@ Other resources:
 - [Angular Awesome resource list](https://github.com/PatrickJS/awesome-angular)
 
```
```suggestion

These are our reference resources. Still, a project often needs extra libraries. Before adding a new library, it must be approved by the project's tech lead.

```

davidgg reviewed (changes requested):
Added agreed changes to the PR.

davidgg commented:
Can you rebase and merge the PR @karlos1337 ?

karlos1337 commented:
Sure!
